packages/docregistry/src/context.ts

```diff
@@ -502,7 +496,38 @@ export class Context<
   /**
    * Handle an initial population.
    */
-  private _populate(): Promise<void> {
+  private async _populate(): Promise<void> {
+    if (this._docProviderFactory) {
+      const serverSettings = ServerConnection.makeSettings();
+      const url = URLExt.join(
+        serverSettings.baseUrl,
+        FILE_PATH_TO_ROOM_ID_URL,
+        encodeURIComponent(this._path)
+      );
+      const data = {
+        format: this._factory.fileFormat!,
+        type: this._factory.contentType
+      };
+      const response = await ServerConnection.makeRequest(
+        url,
+        {
+          method: 'PUT',
+          body: JSON.stringify(data)
+        },
+        serverSettings
+      );
+      if (response.status !== 200 && response.status !== 201) {
+        const err = await ServerConnection.ResponseError.create(response);
+        throw err;
+      }
+      const roomId = await response.text();
+      this._provider = this._docProviderFactory({
+        path: roomId,
+        contentType: this._factory.contentType,
+        format: this._factory.fileFormat!,
+        model: this._model.sharedModel
+      });
+    }
```

Comment:
So this logic should only be enabled for collaborative mode. Since room IDs don't make sense in non-collaborative.

Reply:
I would prefer this logic to be encapsulated within the document provider for two reasons:

- It will respect separation of concern - the request to get a room id is an implementation detailed of the document provider when using the websocket provider. It has no use in the document context.
- It will ease updating/experimenting with other implementation if needed for customer with custom backend infrastructure as the document provider factory is a token:

https://github.com/jupyterlab/jupyterlab/blob/c3c2379421ba2ba0e17cad2e3e3a636af38448fd/packages/docprovider-extension/src/index.ts#L29

> This will require the provider interface to be extended with a `ready: Promise<void>` attribute that resolves when the provider got the room id (in this case). So that we can wait for that promise here in the `_populate` method.

pinging @afshin to get his thought.

Reply:
@fcollonval I think your plan makes sense.